rust-tdshim/src/mp.rs

```diff
@@ -157,55 +159,53 @@ fn parallel_accept_memory(apic_id: u64) {
     }
 }
 
+// Wait for APs to arrive by checking if they are available
 fn wait_for_ap_arrive(cpu_num: u32) {
     let mut mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to arrive...\n");
-    loop {
-        if mail_box.cpu_arrival() == cpu_num - 1 {
-            log::info!("All APs has arrived\n");
-            break;
-        }
-    }
-    mail_box.set_cpu_exiting(cpu_num);
-}
-
-fn wait_for_ap_exit(cpu_num: u32) {
-    let mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to exit...\n");
-    loop {
-        let cpu_exit = mail_box.cpu_exiting();
-
-        if cpu_exit == 1 {
-            log::info!("All APs has exited\n");
-            break;
+    for i in 1..cpu_num {
+        mail_box.set_apic_id(i);
+        mail_box.set_command(MP_WAKEUP_COMMAND_AVAILABLE);
+
+        loop {
+            if mail_box.apic_id() == MAILBOX_APICID_INVALID {
+                mail_box.set_command(MP_WAKEUP_COMMAND_NOOP);
+                break;
+            } else {
+                cpu_pause();
+            }
         }
     }
+    x86::fence::mfence();
 }
 
 pub fn ap_assign_work(apic_id: u32, stack: u64, entry: u32) {
     let mut mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
     mail_box.set_wakeup_vector(entry);
     mail_box.set_fw_arg(0, stack);
 
-    mail_box.set_apid_id(apic_id);
+    mail_box.set_apic_id(apic_id);
     mail_box.set_command(MP_WAKEUP_COMMAND_ACCEPT_PAGES);
 
     loop {
-        let wakeup_apic_id = mail_box.apid_id();
-        if wakeup_apic_id == MAILBOX_APICID_INVALID {
+        if mail_box.apic_id() == MAILBOX_APICID_INVALID {
+            mail_box.set_command(MP_WAKEUP_COMMAND_NOOP);
             log::info!("Successfully wakeup AP #{}\n", apic_id);
             break;
+        } else {
+            cpu_pause();
         }
     }
 }
 
-pub fn mp_accept_memory_resource_range(cpu_num: u32, address: u64, size: u64) {
+pub fn mp_accept_memory_resource_range(mut cpu_num: u32, address: u64, size: u64) {
     log::info!(
         "mp_accept_memory_resource_range: 0x{:x} - 0x{:x} ... (wait for seconds)\n",
         address,
         size
     );
 
+    cpu_num = if cpu_num > 16 { 16 } else { cpu_num };
```

Comment:
Why are we limiting the cpu count to 16? What if my system has >16? If this is some sort of internal limit, please create a top-level `const` with details of this restriction (including links to canonical reference material if possible).

Reply:
This variable refers to the cpu count that will be used to do the memory acceptance, the stacks used for application processors (APs) to run rust function are allocated from temp heap of td-shim which is limited. Besides, it is not necessary to use all the APs to do this work if there's too many.
Will create a `const `and explain the reason, thanks.